File: bootstrap/rb/lib/acute/block.rb
# Acute Language
# Copyright © 2011, Jeremy Tregunna, All Rights Reserved.
#

module Acute
  class Block < ::Acute::Object
    # Blocks have a list of argument names, a Call object, a scope, and a body message.
    attr_accessor :argument_names, :scope, :body
    attr_reader :locals

    def initialize(scope, body, args)
      super()
      @scope = scope
      @body = body
      @locals = ::Acute::Nil.instance
      @argument_names = args.map(&:name) rescue [] #args[0].map(&:name) rescue []
      register(:parent, $state.find("Object"))
      method_table
    end

    def method_table
      method(:call)          { |env| env[:target].call(env) }#, env[:msg].arguments) }
      method(:activate)      { |env| activate(env) }
      method(:message)       { |env| env[:target].body }
      method(:argumentNames) { |env| ::Acute::List.new(env[:target].argument_names.map { |a| ::Acute::String.new(a) }) }
      method(:scope)         { |env| env[:target].scope }
      method(:setScope)      { |env| env[:target].scope = env[:msg].eval_arg_at(env, 0); env[:target] }
    end

    def activate(env)
      call(env)#, env[:msg].arguments)
    end

    def call(env)#, args = [])
      #self.scope = $state.find("Object") unless scope
      create_locals(env, scope)
      argument_names.each_with_index do |name, idx|
        obj = env[:msg].eval_arg_at(env, idx)
        locals.register(name, obj)
      end
      body.perform_on(env, locals, locals)
      #doMessage(env.merge(:msg => body, :target => locals), locals)
    end

    def scope=(other)
      @scope = other
      locals.register(:parent, other)
    end

    private

    def create_locals(env, parent)
      @locals = ::Acute::Object.new
      locals.register(:parent, parent)
      locals.register(:self, scope || env[:target])
      call = ::Acute::Object.new
      call.register(:parent, $state.find("Object"))
      %w{target message sender}.each { |str| call.register(str, env[str.to_sym]) }
      call.register(:slotContext, env[:slot_context])
      call.register(:activated, self)
      locals.register(:call, call)
    end
  end
end

File: wee/block.c
/*
 * Acute Programming Language
 * Copyright (c) 2011, Jeremy Tregunna, All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "block.h"
#include "str.h"
#include "object_space.h"

block_t* block_new(msg_t* body, list_t* names, obj_t* scope)
{
    block_t* blk = (block_t*)obj_new_with_size(sizeof(*blk));
    hash_insert(blk->slots, "message", body);
    list_t* arg_names = list_new();
    list_foreach(names, ^(size_t index, void* element) {
        char* name = (char*)element;
        str_t* str = str_new(name, strlen(name) + 1);
        list_append(arg_names, str);
    });
    hash_insert(blk->slots, "argumentNames", arg_names);
    hash_insert(blk->slots, "scope", scope);

    return blk;
}

void block_destroy(block_t* blk)
{
    msg_t* msg = (msg_t*)hash_get(blk->slots, "message");
    if(msg)
        msg_destroy(msg);
    list_t* arg_names = (list_t*)hash_get(blk->slots, "argumentNames");
    if(arg_names)
        list_destroy(arg_names);
    obj_t* scope = (obj_t*)hash_get(blk->slots, "scope");
    if(scope)
        free(scope);

    free(blk);
    blk = NULL;
}

static obj_t* block_locals_new(void)
{
    obj_t* locals = obj_new();
    return locals;
}

obj_t* block_activate(block_t* blk, obj_t* target, obj_t* sender, msg_t* msg, obj_t* slot_context)
{
    list_t* arg_names = (list_t*)hash_get(blk->slots, "argumentNames");
    obj_t* scope = (obj_t*)hash_get(blk->slots, "scope");
    if(scope == NULL)
        scope = target;

    obj_t* locals = block_locals_new();

    if(arg_names)
    {
        list_foreach(arg_names, ^(size_t index, void* element) {
            obj_t* val = msg_eval_arg_at(msg, sender, index);
            obj_register_slot(locals, (char*)element, val);
        });
    }

    msg_t* message = (msg_t*)hash_get(blk->slots, "message");
    return msg_perform_on(message, locals, locals);
}

obj_t* block_call(block_t* blk, obj_t* locals, msg_t* msg)
{
    return block_activate(blk, locals, locals, msg, locals);
}